### ISSUE TITLE
Consider using `.after()` inside `.addHook` and `.use` calls

### ISSUE THREAD

**Issue by nwoltman** [fastify/fastify#748]
Here's a summary of the bug from fastify/fastify-cookie#5:

```js
fastify.register(require('fastify-cookie'))

fastify.addHook('preHandler', (request, reply, next) => {
  console.log(request.cookies) // {} - Expected cookies to have been parsed here :(
  next()
})

fastify.get('/', (request, reply) => {
  console.log(request.cookies) // {foo: 'bar'} - Now the cookies have been parsed
  reply.send('hello')
})
```

Since plugins are registered asynchronously, the user's call to `fastify.addHook` happens before `fastify-cookie` calls `.addHook`, so even though it looks like the user added their hook after `fastify-cookie`, it actually gets added before the cookie hook.

This could be fixed by using `this.after()` inside `.addHook` so that the hook would get added after the plugin's hooks. We already do this for `fastify.route` (which includes the route shorthands) as well as `fastify.setNotFoundHandler`.

Note that this change would affect code like this:

```js
fastify.addHook('preHandler', function preHandlerA () {})

fastify.register((instance, options, next) => {
  instance.get('/user', function handler () {})
  // Currently: Both preHandlerA and preHandlerB run before the '/user' handler
  // With .after(): Only preHandlerA runs before the '/user' handler
})

fastify.addHook('preHandler', function preHandlerB () {})
```

I think that using `.after()` inside `.addHook` and `.use` calls would be a good idea because it would make the ordering of hooks and middleware more predictable. Does everyone agree?

**Comment by mcollina** [comment]
I think this is a good idea. @delvedor what do you think?

**Comment by jsumners** [comment]
Makes sense to me.

**Comment by allevo** [comment]
It is ok for me.

**NB:** In the "hook pattern", each listener has to be independent from each other.

**Comment by nwoltman** [comment]
I've been trying to implement this and now I have a question.

Since this change means that handlers—and not-found handlers—will only run hooks/middleware that are defined before the handler, what do we do about the default not-found handler? Currently it gets set when the `fastify` instance is created, which means that it won't run any hooks or middleware. Is that okay? Or should we make it a special case and allow all non-encapsulated hooks and middleware to run before it?

The decision on this would solve this current issue:

```js
const fastify = require('fastify')()

fastify.register(require('a-plugin-with-a-preHandler'))

fastify.addHook('preHandler', function preHandlerA (request, reply, next) {
  request.preHandlerA = true
  next()
})

fastify.listen()
```

The default not-found handler will have `preHandlerA` run before it, but not the `preHandler` from the plugin.

**Comment by mcollina** [comment]
> Since this change means that handlers—and not-found handlers—will only run hooks/middleware that are defined before the handler, what do we do about the default not-found handler? Currently it gets set when the fastify instance is created, which means that it won't run any hooks or middleware. Is that okay? Or should we make it a special case and allow all non-encapsulated hooks and middleware to run before it?

I don't think this should be the correct plan, ideally all hooks and middlewares defined "at the same level" should all be executed. I think we might want to keep things the way we are for 1.0.0, and rethink the encapsulation model a bit for 2.0.0.

**Comment by nwoltman** [comment]
In that case, we'll need to document the `fastify.after()` method.

**Comment by nwoltman** [comment]
Although a problem currently is that the levels aren't consistent. For example:

```js
const fp = require('fastify-plugin')
const fastify = require('fastify')()

fastify.decorateRequest('payload', '')

fastify.register(fp(async function (instance) {
  instance.addHook('preHandler', async function A (request) { // third
    request.payload += 'A'
  })
}))

fastify.addHook('preHandler', async function B (request) { // first
  request.payload += 'B'
})

fastify.get('/', function handler (request, reply) {
  reply.send(request.payload)
})

fastify.register(fp(async function (instance) {
  instance.addHook('preHandler', async function C (request) { // never runs
    request.payload += 'C'
  })
}))

fastify.addHook('preHandler', async function D (request) { // second
  request.payload += 'D'
})

fastify.inject('/', (err, res) => {
  console.log(res.payload) // 'BDA'
})
```

We'll need to explain in the documentation to always register plugins before routes that need them and to use `.after()` for hooks/middleware that need to come after plugins.

**Comment by delvedor** [comment]
I agree that we should update the docs and make a recommendation on how register the components of the application.

```
plugins (cookies, compression...)
decorators
routes (could be also a plugin)
```

I'll work on this.
Regarding the internal usage of `after` to register hooks and middlewares I'm 👍 .

**Comment by nwoltman** [comment]
@delvedor If we internally use `after` when registering hooks and middleware then the result of my example will be `AB`. I think this would be a better developer experience because linear ordering is easier to work with (it's also how Express and Koa register middleware), but I'm wondering if you agree?

**Comment by mcollina** [comment]
I think we might want to improve a bit the situation. Instead of generating an hook and middleware runner for every route, we should generate them for every plugin, after the plugin has finished loading. This should solve most of the issues. I remember we implemented them in this way in the beginning, but I do not remember why we changed them.

**Comment by delvedor** [comment]
@mcollina to do what you suggest we should figure out a way to understand when a plugin has finished to load (can we do that outside avvio?) and then register hooks - middlewares - routes (in this order).
Basically we should popolate the context object and register the routes only at the end of the plugin work (how will this work top level?).

We add hook and middlewares for every route because we must store them inside the [store](https://github.com/delvedor/find-my-way#onmethod-path-handler-store) object of `find-my-way`.

**Comment by mcollina** [comment]
> Basically we should popolate the context object and register the routes only at the end of the plugin work (how will this work top level?).

No. I mean that the route context links to a plugin context, which includes all those details which will be shared.

**Comment by nwoltman** [comment]
@mcollina That makes sense. It was probably changed because of this:

```js
fastify.register(require('plugin-1'))

fastify.get('/1', ...) // has only plugin-1

fastify.register(require('plugin-2'))

fastify.get('/2', ...) // has plugin-1 and plugin-2 */)
```

They're all at the same level, but `plugin-2` gets registered after the `get('/1')` route is already set up.

**Comment by nwoltman** [comment]
@mcollina A solution would be to have hooks/middlware use `after` and then have routes use `ready`. That way, routes will be set up once all of the plugins/hooks/middleware for the Fastify instance are ready. However, this is only possible if `ready` doesn't automatically start the boot process.

**Comment by mcollina** [comment]
Let’s just document what we have and ship 1.0.0. Then we can focus on this issue

**Comment by mcollina** [comment]
This was regressed in https://github.com/fastify/fastify/pull/474. This is caused by the hookRunner (and before that fast-iterator) creating a closure inside `.after` from within `.route`. Note that this happens only when plugins are involved, otherwise after takes care of them.

This is a test that reproduces the problem:

```
diff --git a/test/hooks.test.js b/test/hooks.test.js
index c0f00e6..6db8f38 100644
--- a/test/hooks.test.js
+++ b/test/hooks.test.js
@@ -6,6 +6,7 @@ const sget = require('simple-get').concat
 const stream = require('stream')
 const Fastify = require('..')
 const fs = require('fs')
+const fp = require('fastify-plugin')

 const payload = { hello: 'world' }

@@ -1252,3 +1253,44 @@ if (Number(process.versions.node[0]) >= 8) {
   t.pass('Skip because Node version < 8')
   t.end()
 }
+
+test('Register an hook after a plugin', t => {
+  t.plan(6)
+  const fastify = Fastify()
+
+  fastify.register(fp(function (instance, opts, next) {
+    instance.addHook('preHandler', function (req, reply, next) {
+      t.ok('called')
+      next()
+    })
+
+    instance.get('/', function (request, reply) {
+      reply.send({ hello: 'world' })
+    })
+
+    next()
+  }))
+
+  fastify.register(fp(function (instance, opts, next) {
+    fastify.addHook('preHandler', function (req, reply, next) {
+      t.ok('called')
+      next()
+    })
+
+    fastify.addHook('preHandler', function (req, reply, next) {
+      t.ok('called')
+      next()
+    })
+
+    next()
+  }))
+
+  fastify.inject({
+    url: '/',
+    method: 'GET'
+  }, (err, res) => {
+    t.error(err)
+    t.is(res.statusCode, 200)
+    t.deepEqual(JSON.parse(res.payload), { hello: 'world' })
+  })
+})
```

**Comment by delvedor** [comment]
The problem is generated here:
https://github.com/fastify/fastify/blob/efc4525236b1d2831521fd3735001d110efc83dd/fastify.js#L490

Since we use concat we loose the reference to the instance hooks and we don't get any further updates. If you run the same test with `onRequest` everything will work as expected.
I'll work on a solution.

**Comment by delvedor** [comment]
Closed by #762.

**Comment by nwoltman** [comment]
With the fix from #762 the result of my [example](#issuecomment-362954220) will be `BDAC`. @delvedor are you no longer interested in using `after` to make the result `ABCD`?

**Comment by delvedor** [comment]
Currently I have no time for that, but if you can, fire a pr! :)
